### PR TITLE
Implement all different resolutions

### LINKED ISSUE #344: Change the `compute_rp_partition` function strategies
### What and Why

Currently, we have two strategies in the `compute_rp_partition`: `:greedy, :all`

We use the `:greedy` strategy to determine the lowest resolution for the balance constraints, whereas the `:all` strategy determines capacity constraint limits.

The `:all` strategy considers the outgoing and incoming flows to determine the highest resolution. This creates redundant constraints in the maximum outflow/inflow capacity limit. So, we need strategies for each one individually to have the highest resolution for the output flows and the highest resolution for the input flows separately.

We are not going to use the current `:all` strategy for the time being. Maybe we will use it in the future, I'm not sure. @g-moralesespana what do you think? any extra thoughts on this issue?

### Possible Drawbacks

More pre-computed partitions can lead to higher building model times (or maybe not...since there are more partitions to calculate but much simpler to determine)

### Related Issues

#355

### PR DESCRIPTION
**Pull request by abelsiqueira**
# Pull request details

## Describe the changes made in this pull request

Implement the various constraints resolutions.

## List of related issues or pull requests

Closes #344 

## Collaboration confirmation

As a contributor I confirm

- [x] I read and followed the instructions in README.dev.md
- [x] The documentation is up to date with the changes introduced in this Pull Request (or NA)
- [x] Tests are passing
- [x] Lint is passing

### CODE REVIEW

**Comment by abelsiqueira** [comment]
@datejada, do you see anything obviously wrong in these short 15k lines: https://gist.github.com/abelsiqueira/91889edd1b9b065dbe396e355185a9e1

**Comment by abelsiqueira** [comment]
This is the feasibility report:

```
  consumer_balance[demand,1,23:23] : flow[(ocgt, demand), 1, 23:23] + flow[(ccgt, demand), 1, 23:23] + flow[(wind, demand), 1, 23:23] + flow[(solar, demand), 1, 23:23] … => 2.27374e-13
  consumer_balance[demand,3,14:14] : flow[(ocgt, demand), 3, 14:14] + flow[(ccgt, demand), 3, 14:14] + flow[(wind, demand), 3, 14:14] + flow[(solar, demand), 3, 14:14] … => 1.13687e-13
  consumer_balance[demand,1,8:8] : flow[(ocgt, demand), 1, 8:8] + flow[(ccgt, demand), 1, 8:8] + flow[(wind, demand), 1, 8:8] + flow[(solar, demand), 1, 8:8] + flow[(en… => 1.13687e-13
  consumer_balance[demand,1,12:12] : flow[(ocgt, demand), 1, 12:12] + flow[(ccgt, demand), 1, 12:12] + flow[(wind, demand), 1, 12:12] + flow[(solar, demand), 1, 12:12] … => 1.13687e-13
  consumer_balance[demand,2,11:11] : flow[(ocgt, demand), 2, 11:11] + flow[(ccgt, demand), 2, 11:11] + flow[(wind, demand), 2, 11:11] + flow[(solar, demand), 2, 11:11] … => 1.13687e-13
  consumer_balance[demand,1,20:20] : flow[(ocgt, demand), 1, 20:20] + flow[(ccgt, demand), 1, 20:20] + flow[(wind, demand), 1, 20:20] + flow[(solar, demand), 1, 20:20] … => 2.27374e-13
  consumer_balance[demand,1,15:15] : flow[(ocgt, demand), 1, 15:15] + flow[(ccgt, demand), 1, 15:15] + flow[(wind, demand), 1, 15:15] + flow[(solar, demand), 1, 15:15] … => 1.13687e-13
  consumer_balance[demand,1,21:21] : flow[(ocgt, demand), 1, 21:21] + flow[(ccgt, demand), 1, 21:21] + flow[(wind, demand), 1, 21:21] + flow[(solar, demand), 1, 21:21] … => 2.27374e-13
  ```

**Comment by abelsiqueira** [comment]
@datejada, now a more reasonable comparison: https://gist.github.com/abelsiqueira/92c975ec412848246ffd7ac4545f9017

Old and New are different, because of the resolution of the hub and conversion. And possibly something else.
I used https://meldmerge.org to compare these two files more easily.

**Comment by datejada** [comment]
@abelsiqueira I expect it to be different since we are changing the strategies. For instance, the hub was previously defined in the lowest resolution and will now be in the highest. So, let me check it with @g-moralesespana tomorrow in the office and let you know an answer.

BTW: Can this branch be in the main repo to do a benchmark? Since we are creating more constraints, I would like to see the impact on the performance. Thanks!

**Comment by abelsiqueira** [comment]
When I manage to fix it I will push to the main repo.

**Comment by datejada** [comment]
Hi @abelsiqueira, I have reviewed the LP files. The new one is according to the table in #344. So, the changes in the hub balance, conversion balance, and the phs output and input limits are expected 😄 

This means that the objective function in the Tiny case study with the variable resolution must change. The Norse objective function will also change. But, the Tiny and the EU benchmark cases cannot change the objective functions compared to the existing ones.

German is on holiday for the rest of the week, but we can show him the LP on Monday.

Thanks!

**Comment by abelsiqueira** [comment]
But it's infeasible for the variable resolution case. So something is wrong, no? I am (trying) to finalise the aggregation changes as well, but if I did it right, it is still infeasible!?
Anyway, some exclamation marks for drama, but the question is whether the model has some issue or the code has some issue.

**Comment by abelsiqueira** [comment]
In particular, as an example, what should be the constraints `max_output_flows_limit` for wind?

**Comment by datejada** [comment]
Ah! ok, sorry, I didn't understand that it was infeasible 😞 
So, it is more challenging then. Let me see it in more detail and come back to you. It will take more time than I thought

**Comment by datejada** [comment]
@abelsiqueira The problem is that this constraint shouldn't have the duration multiplying the transport flow (or you multiply the duration in the RHS).

max_transport_flow_limit[(balance,demand),1,1:3] : 3 flow[(balance, demand), 1, 1:3] <= 200

In the old file was:
max_transport_flow_limit_(balance,demand),1,1_3_: 3 flow_(balance,_demand),_1,_1_3_ <= 600

Fixing this will make it feasible 😄

**Comment by datejada** [comment]
@abelsiqueira Sorry, I might not have been totally clear in my previous comment. The equation that we expect is the following:

max_transport_flow_limit[(balance,demand),1,1:3] : flow[(balance, demand), 1, 1:3] <= 200

The thing is that in the previous version, it was ok because it is equivalent to the one in this comment. Only because the durations on both sides were cancelled.

I hope this comment make it more clear 😄

**Comment by abelsiqueira** [comment]
I've updated the model but it is still infeasible. The latest version is new.lp on https://gist.github.com/abelsiqueira/92c975ec412848246ffd7ac4545f9017
There are more changes than just this last duration change. I also change the aggregation of the flows for some cases - When you have `power`, right?
Anyway, I will make the PR cleaner soon, so we can discuss it Monday if we can't find anything today.

**Comment by datejada** [comment]
@abelsiqueira, The implementation is according to what we defined, so thanks for that.

We need to review what we defined @g-moralesespana on Monday, especially the conversion balance. By doing the conversion balance in power, the problem can become infeasible because it is too restrictive (and it might be wrong?), and the conversion balance would make more sense to do it in energy, but let's discuss it.

These are the constraints that are making it infeasible:

```
 consumer_balance_demand,1,1_3_ : flow_(balance,_demand),_1,_1_3_ == 85
 consumer_balance_demand,1,4_6_ : flow_(balance,_demand),_1,_4_6_ == 69.99999999999999
 hub_balance_balance,1,3_3_ : flow_(ccgt,_balance),_1,_3_3_ + flow_(wind,_balance),_1,_3_6_ + flow_(phs,_balance),_1,_1_4_ - flow_(balance,_demand),_1,_1_3_ == 0
 hub_balance_balance,1,4_4_ : flow_(ccgt,_balance),_1,_4_4_ + flow_(wind,_balance),_1,_3_6_ + flow_(phs,_balance),_1,_1_4_ - flow_(balance,_demand),_1,_4_6_ == 0
 conversion_balance_ccgt,1,3_3_ : flow_(H2,_ccgt),_1,_1_6_ - flow_(ccgt,_balance),_1,_3_3_ == 0
 conversion_balance_ccgt,1,4_4_ : flow_(H2,_ccgt),_1,_1_6_ - flow_(ccgt,_balance),_1,_4_4_ == 0
```

If one starts replacing the constraints, the resulting constraints in the hub balance are:

```
 hub_balance_balance,1,3_3_ : flow_(H2,_ccgt),_1,_1_6_ + flow_(wind,_balance),_1,_3_6_ + flow_(phs,_balance),_1,_1_4_  == 85
 hub_balance_balance,1,4_4_ : flow_(H2,_ccgt),_1,_1_6_ + flow_(wind,_balance),_1,_3_6_ + flow_(phs,_balance),_1,_1_4_  == 69.99999999999999
```

Which clearly is infeasible 😞 

Having the conversion balance in energy and keeping the output limit in power is the solution.

**Comment by datejada** [comment]
@abelsiqueira, small fix: would you mind adding a name to the storage level max limit constraint? currently, it doesn't have a name, so it appears as cXX in the lp file. Thanks!

**Comment by abelsiqueira** [comment]
Done

File: src/create-model.jl [hunk 1]
```diff
@@ -283,35 +284,77 @@ function create_model(
...
     # Below, we drop zero coefficients, but probably we don't have any
     # (if the implementation is correct)
-    drop_zeros!.(incoming_flow_lowest_resolution)
-    drop_zeros!.(outgoing_flow_lowest_resolution)
-    drop_zeros!.(incoming_flow_highest_resolution)
-    drop_zeros!.(outgoing_flow_highest_resolution)
+    # drop_zeros!.(incoming_flow_lowest_resolution)
+    # drop_zeros!.(outgoing_flow_lowest_resolution)
+    # drop_zeros!.(incoming_flow_highest_resolution)
+    # drop_zeros!.(outgoing_flow_highest_resolution)
```
Comment (datejada) [inline src/create-model.jl:357]:
```suggestion
    drop_zeros!.(incoming_flow_lowest_resolution)
    drop_zeros!.(outgoing_flow_lowest_resolution)
    drop_zeros!.(incoming_flow_highest_resolution)
    drop_zeros!.(outgoing_flow_highest_resolution)
```

**Review by datejada — APPROVED** [review]
I'm approving since the suggested changes are minor. Thanks!

**Pull request merged**